### PR TITLE
[refactor] Refactor normalizers and encoders

### PR DESCRIPTION
**Pull request by ervteng**
### Proposed change(s)

Refactor the NetworkBody, Normalizer, and Encoders as specified in the design: 

https://docs.google.com/document/d/11Spnb99cJr_w2BYSwQBVcGAepeWBgQR0nj189yupAc0/edit#

TODO: ActorCritic and SeparateActorCritic

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/torch/encoders.py [hunk 1]
```diff
@@ -66,14 +57,78 @@ def conv_output_shape(h_w, kernel_size=1, stride=1, pad=0, dilation=1):
...
+        hidden_size: int,
+        num_actions: int,
+        num_layers: int,
+        normalize: bool = False,
+    ):
+        super().__init__(
+            input_size + num_actions, hidden_size, num_layers, normalize=False
+        )
+        if normalize:
+            self.normalizer = Normalizer(input_size)
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/encoders.py:112]: This is not really an Action Vector Encoder, it also encodes "inputs" and concatenates it with the actions. 
I also don't see the use case for a normalizer for the inputs here. Are the inputs expected to be always vector observations?

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 2]
```diff
@@ -56,8 +56,9 @@ def create_encoders(
         h_size: int,
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:56]: This is only for vector observations right? I wonder if this create_encoders should create a hidden layer for vector_observations. I think most of the time, we simply concatenate the normalized vector_observations (without an additional layer) with the visual observations encodings.

File: ml-agents/mlagents/trainers/torch/utils.py [hunk 3]
```diff
@@ -56,8 +56,9 @@ def create_encoders(
...
         num_layers: int,
         vis_encode_type: EncoderType,
-        action_size: int = 0,
-    ) -> Tuple[nn.ModuleList, nn.ModuleList, nn.ModuleList]:
+        encoded_act_size: int = 0,
+        normalize: bool = False,
+    ) -> Tuple[nn.ModuleList, nn.ModuleList]:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/utils.py:61]: Modify the docstring here, the inputs and outputs of this method changed sllightly

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:56]: I'm pretty sure in TF we add some hidden layers to the vector_observations before concatenating with visual observations - or do you mean the visual_observations? I did find a bug where we're not adding any hidden layers to the visual obs in Torch.

File: ml-agents/mlagents/trainers/torch/encoders.py [hunk 4]
```diff
@@ -66,14 +57,78 @@ def conv_output_shape(h_w, kernel_size=1, stride=1, pad=0, dilation=1):
     return h, w
 
 
-def pool_out_shape(h_w, kernel_size):
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/torch/encoders.py:69]: Should we add `pool_out_shape` and `conv_output_shape` to the torch utils file?

File: ml-agents/mlagents/trainers/torch/encoders.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/encoders.py:112]: Renamed to `VectorAndUnnormalizedInputEncoder` - how's that?

File: ml-agents/mlagents/trainers/torch/utils.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/torch/utils.py:61]: Updated
